File: LeetCode/python3/731.py
class MyCalendarTwo:
    calendar = []
    doubleCalendar = []

    def __init__(self):
        self.calendar = []
        self.doubleCalendar = []

    def book(self, start, end):
        """
        :type start: int
        :type end: int
        :rtype: bool
        """
        for i, j in self.doubleCalendar:
            if end > i and start < j:
                return False
        for i,j in self.calendar:
            if end > i and start < j:
                self.doubleCalendar.append([max(start, i), min(end, j)])
        self.calendar.append([start, end])
        return True

File: LeetCode/cpp/731.cpp
class MyCalendarTwo {
public:
    std::vector<std::vector<int>> calendar, doubleCalendar;

    MyCalendarTwo() {
        calendar = std::vector<std::vector<int>>();
        doubleCalendar = std::vector<std::vector<int>>();
    }

    bool book(int start, int end) {
        for (auto iter:doubleCalendar) {
            if (start < iter[1] && end > iter[0])
                return false;
        }
        for (auto iter:calendar) {
            if (end > iter[0] && start < iter[1])
                doubleCalendar.push_back({std::max(start, iter[0]), std::min(end, iter[1])});
        }
        calendar.push_back({start, end});
        return true;
    }
};
